emscripten.py

```diff
@@ -37,14 +37,46 @@
 WASM_INIT_FUNC = '__wasm_call_ctors'
 
 
+def make_wasm_table_static_dyncaller(func):
+  sig = func.replace('dynCall_', '')
+  ret = '' if sig[0] == 'v' else 'return '
+  args = '' # 'a0, a1, a2, ..., aN'
+  ptr_args = '' # 'ptr, a0, a1, a2, ..., aN'
+  i = 0
+  while i < len(sig)-1:
+    if i > 0: args += ', '
+    args += 'a' + str(i)
+    i += 1
+  ptr_args = ('ptr, ' + args) if len(args) > 0 else 'ptr'
+
+  dyncall = ('dyncalls["' + sig + '"]') if shared.Settings.MINIMAL_RUNTIME else ('Module["' + func + '"]')
+  wasmTableGet = 'wasmTableGet' if shared.Settings.SHRINK_LEVEL == 0 else 'wasmTable.get'
+  return 'function ' + func + '(' + ptr_args + ') { ' + ret + dyncall + '(' + ptr_args + '); }\n'
+
+
 def compute_minimal_runtime_initializer_and_exports(post, exports, receiving):
   # Declare all exports out to global JS scope so that JS library functions can access them in a
   # way that minifies well with Closure
   # e.g. var a,b,c,d,e,f;
   exports_that_are_not_initializers = [x for x in exports if x not in WASM_INIT_FUNC]
   # In Wasm backend the exports are still unmangled at this point, so mangle the names here
   exports_that_are_not_initializers = [asmjs_mangle(x) for x in exports_that_are_not_initializers]
+
+  static_dyncall_sig_functions = ''
+
+  if shared.Settings.WASM_DYNCALLS:
+    if len([x for x in exports_that_are_not_initializers if x.startswith('dynCall_')]) > 0:
+      exports_that_are_not_initializers += ['dynCalls = {}']
+  else:
+    for x in exports_that_are_not_initializers:
+      if x.startswith('dynCall_'):
```

Comment:
I think this assumes that `dynCall_` functions are exported from the wasm? That isn't the case for `WASM_BIGINT` or otherwise when we do not legalize, and is definitely not the case when `wasm-emscripten-finalize` does not write out any changes at all to the wasm (then the wasm is identical to what LLVM emitted).

I think we need to make `wasm-emscripten-finalize` print out the list of function pointer types in the metadata.